appinterface/projection/crossfire/crossfire.go

```diff
@@ -0,0 +1,754 @@
+				JoinedAtBlockHeight:             blockHeight,
+				JoinedAtBlockTime:               blockTime,
+				Moniker:                         msgCreateValidatorEvent.Description.Moniker,
+				Identity:                        msgCreateValidatorEvent.Description.Identity,
+				Website:                         msgCreateValidatorEvent.Description.Website,
+				SecurityContact:                 msgCreateValidatorEvent.Description.SecurityContact,
+				Details:                         msgCreateValidatorEvent.Description.Details,
+				TaskPhase1NodeSetup:             constants.INCOMPLETED,
+				TaskPhase2KeepNodeActive:        constants.INCOMPLETED,
+				TaskPhase2ProposalVote:          constants.INCOMPLETED,
+				TaskPhase2NetworkUpgrade:        constants.INCOMPLETED,
+				RankTaskPhase1n2CommitmentCount: 0,
+				RankTaskPhase3CommitmentCount:   0,
+				RankTaskHighestTxSent:           0,
+			}
+
+			isJoined, joinedAtBlockHeight, joinedAtBlockTime, err := crossfireValidatorsView.LastJoinedBlockHeight(
+				validatorRow.OperatorAddress, validatorRow.ConsensusNodeAddress,
+			)
+			if err != nil {
+				return fmt.Errorf("error querying validator last joined block height: %v", err)
+			}
+			if isJoined {
+				validatorRow.JoinedAtBlockHeight = joinedAtBlockHeight
+				validatorRow.JoinedAtBlockTime = joinedAtBlockTime
+			}
+
+			if err := crossfireValidatorsView.Upsert(&validatorRow); err != nil {
+				return fmt.Errorf("error inserting new validator into view: %v", err)
+			}
+
+			// checkTaskSetup
+			if err := projection.checkTaskSetup(
+				crossfireValidatorsView,
+				validatorRow.OperatorAddress,
+				validatorRow.ConsensusNodeAddress,
+				validatorRow.JoinedAtBlockTime,
+			); err != nil {
+				return fmt.Errorf("error check Setup task for new validator: %v", err)
+			}
+		}
+	}
+	for _, event := range events {
+		if msgSubmitSoftwareUpgradeProposalEvent, ok := event.(*event_usecase.MsgSubmitSoftwareUpgradeProposal); ok {
+			projection.logger.Debug("handling MsgSubmitSoftwareUpgradeProposal event")
+
+			// Check if proposed after competition has ended
+			if blockTime.After(projection.competitionEndTime) {
+				return fmt.Errorf("error Competition has already ended")
+			}
+
+			// Check if proposed before OR after Phase 2
+			if blockTime.Before(projection.phaseTwoStartTime) || blockTime.After(projection.phaseThreeStartTime) {
+				return fmt.Errorf("error This proposal does not occur in Phase 2")
+			}
+
+			// Check if proposed by NOT an admin
+			if msgSubmitSoftwareUpgradeProposalEvent.ProposerAddress != projection.adminAddress {
+				return fmt.Errorf("error checking proposer address equals admin address")
+			}
+
+			// Check if proposal ID does not match the required ID
+			if msgSubmitSoftwareUpgradeProposalEvent.MaybeProposalId != nil && *msgSubmitSoftwareUpgradeProposalEvent.MaybeProposalId != projection.networkUpgradeProposalID {
+				return fmt.Errorf("error checking Proposal ID in proposal")
```

Comment:
It is possible that the proposal is proposed by someone else. Here should just ignore it.